### ISSUE TITLE
headline-node-property drawer is inserted after a newline (i.e. invalid syntax)

### ISSUE THREAD

**Issue by hrehfeld** [ndwarshuis/om.el#22]
Here's a test case. Evaluate the following elisp while on the `foo` headline (e.g. with `eval-expression`):

``` emacs-lisp
(let ((headline (org-ml-parse-this-subtree)))
	(save-excursion
		(goto-char (point-min))
		(insert (org-ml-to-string (org-ml-headline-set-node-property "FOO" "BAR" headline)))))
```

``` org-mode
* foo

** bar
```

With org-ml Version: 4.0.1, this inserts on my machine:

``` org-mode
* foo

:PROPERTIES:
:FOO:      BAR
:END:
** bar


```

This (missing newline) works fine:

``` org-mode
* foo
** bar
```

It seems that set-node-property returns a weird section before the property drawer?:
``` lisp
(headline (:raw-value "foo" :begin 1 :end 15 :pre-blank 0 :contents-begin 7 :contents-end 14 ...) 
    (section (:post-blank 0 :begin nil :end nil :parent nil :contents-begin nil :contents-end nil ...)
    (property-drawer (:post-blank 0 :begin nil :end nil :parent nil :contents-begin nil :contents-end nil ...) (node-property ...)))
    (headline (:raw-value "bar" :begin 7 :end 14 :pre-blank 0 :contents-begin nil :contents-end nil ...)))
```

**Comment by ndwarshuis** [comment]
This is supposed to happen. What you are seeing is the `:pre-blank` property on the headline (which corresponds to "spacing before the first child node"). This is done by `org-element`, and why exactly it assigns the space to `:pre-blank` vs `:post-blank` (which corresponds to "space after the node") I'm not sure.

If you wanted to get rid of the pre-blank you would need something like this:

``` emacs
(let ((headline (org-ml-parse-this-subtree)))
    (save-excursion
        (goto-char (point-min))
            (insert (org-ml-to-string (org-ml-set-property :pre-blank 0 (org-ml-headline-set-node-property "FOO" "BAR" headline))))))
```

> It seems that set-node-property returns a weird section before the property drawer?

This is also by design. The section wraps everything underneath the headline that isn't another headline (which in this case is just the property drawer but could have tables, paragraphs, etc).

**Comment by hrehfeld** [comment]
Hmm, I'm not very familiar with the org-element api, but I'm pretty sure that the newline makes the `PROPERTIES` drawer invalid syntax, as it ["has to be located right below a headline"](https://orgmode.org/manual/Property-Syntax.html#Property-Syntax). `org-mode` behaves accordingly, e.g. try inserting a new property with the newline present -- it will insert a new properties drawer above the blank line.

Your suggestion just removes the space (which is a work-around), but I think that space is wrong in the first place (which might very well be an org-element bug). It's either supposed to be pre-blank on the child headline or post-blank on the section of the headline.

Anyways, it's very unexpected that `(org-ml-to-string (org-ml-headline-set-node-property "FOO" "BAR" (org-ml-parse-this-subtree))` doesn't create "valid" (i.e. does what the code says) output. Maybe we need to file an upstream issue?

**Comment by ndwarshuis** [comment]
OK I see now.

### Short story

`org-element` is doing what it is supposed to do and `org-ml`'s method of 'inserting' a property drawer needs to take spacing into account and shift spaces appropriately to conform to valid syntax.

### Long story

This first headline has a `:pre-blank` of 1; in this case it is accurate to use `:pre-blank` and not `:post-blank` of the section because of the "something" string:

```
* foo

something
** bar
```

Adding a property drawer should move `:pre-blank` of the headline to the `:post-blank` of the new drawer (which I think is what you were getting at):

```
* foo
:PROPERTIES:
:FOO: bar
:END:

something
** bar
```

### Even longer story

This isn't just about `:pre-blank` on headlines, because the same thing will happen if you have this and you try to insert a property drawer using `org-ml` (eg there will be a space between the DEADLINE planning node and the drawer, which isn't valid org syntax, in this case the space is from `:post-blank` on the planning node):

```
* foo
DEADLINE: <2020-01-01 Fri>

something
** bar
```

To generalize, the 'metadata' of the headline (a term I just made up to describe the planning, property drawer, and logbook) should all be one continuous wall of text with no spaces, so 'inserting' any of these should move the `:post-blank` (or `:pre-blank` if there is no 'metadata' yet) to the `:post-blank` of whatever is being inserted.

**Comment by ndwarshuis** [comment]
should be fixed with 25ba83621ca0e6653cd466f7983137486350c273, 907e96c1e1fba4ac082d72084d2ceac481746ea7, and 80de56a3ae5c1abeacfe815579b0d63c69a0e7dc
